File: daily/Sliding Window/930-Binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        return self.atMost(nums, goal) - self.atMost(nums, goal - 1)
    def atMost(self, nums: List[int], goal: int) -> int:
        if goal < 0:
            return 0
        left = 0
        n = len(nums)
        currentSum = 0
        counted = 0
        for right in range(n):
            currentSum += nums[right]
            while currentSum > goal:
                currentSum -= nums[left]
                left += 1
            counted += (right - left + 1)
        return counted

File: daily/Sliding Window/930-Binary-subarrays-with-sum.go
package daily

func numSubarraysWithSum(nums []int, goal int) int {
	return atMost(nums, goal) - atMost(nums, goal-1)
}
func atMost(nums []int, goal int) int {
	if goal < 0 {
		return 0
	}
	left := 0
	n := len(nums)
	currentSum := 0
	counted := 0
	for right := 0; right < n; right++ {
		currentSum += nums[right]
		for currentSum > goal {
			currentSum -= nums[left]
			left++
		}
		counted += (right - left + 1)
	}
	return counted
}
